### PR TITLE
io: introducing TBufferIO class

### PR DESCRIPTION
**Pull request by linev**
It is base class for TBufferText and TBufferFile, combining common methods of both.

It implements methods to work with objects map and some other virtual methods, defined in TBuffer interface.

In TBufferFile **nullptr** replaces 0 (where it make sense)  and ``printf()`` replaced with ``Info()``
Also clang format applied to the sources (excluding TBufferFile), use

As a next step, one can put following methods in TBufferIO:
   * ReadClassBuffer
   * WriteClassBuffer
   * ReadClassEmulated
   * ApplySequence

But here one need tiny virtual interface to use proper actions list for TBufferText or TBufferFile.

### CODE REVIEW

File: io/io/inc/TBufferFile.h [hunk 1]
```diff
@@ -38,74 +38,40 @@ class TVirtualStreamerInfo;
...
    typedef std::vector<TStreamerInfo*> InfoList_t;
 
-   Int_t           fMapCount;      ///< Number of objects or classes in map
-   Int_t           fMapSize;       ///< Default size of map
-   Int_t           fDisplacement;  ///< Value to be added to the map offsets
-   UShort_t        fPidOffset;     ///< Offset to be added to the pid index in this key/buffer.
-   TExMap         *fMap;           ///< Map containing object,offset pairs for reading/writing
-   TExMap         *fClassMap;      ///< Map containing object,class pairs for reading
-   TStreamerInfo  *fInfo;          ///< Pointer to TStreamerInfo object writing/reading the buffer
+   TStreamerInfo  *fInfo{nullptr};  ///< Pointer to TStreamerInfo object writing/reading the buffer
```
Comment (pcanal) [inline io/io/inc/TBufferFile.h]: This is used in a few places in the CollectionProxy implementation.  Is it really not needed for the Text versions?

Comment (linev) [reply io/io/inc/TBufferFile.h]: > Is it really not needed for the Text versions?

In text version full stack of streamer infos is preserved.
One should check if **fInfo** in TBufferFile always contains that is expected.
For instance, if one store objects with other objects as data members like:

    class Test {
        int fInt;
        TNamed fNamed;
        Float16_t fFloat;
    };

Question - does **fInfo** really points on Test class after storing **fNamed** member?

Comment (pcanal) [reply io/io/inc/TBufferFile.h]: Maintaining this is the job of IncrementLevel and DecrementLevel ...

Comment (linev) [reply io/io/inc/TBufferFile.h]: > Maintaining this is the job of IncrementLevel and DecrementLevel ...

Now I see in TBufferFile.cxx source code. 
Actually, TBufferJSON/XML/SQL maintain own stack of streamer infos, but with additional information. Therefore I will keep logic as is

**Pull request merged**